modyn/trainer_server/internal/trainer/remote_downsamplers/remote_rho_loss_downsampling.py

```diff
@@ -0,0 +1,69 @@
+from typing import Optional, Union
+
+import torch
+from modyn.trainer_server.internal.trainer.remote_downsamplers.abstract_remote_downsampling_strategy import (
+    AbstractRemoteDownsamplingStrategy,
+)
+from modyn.trainer_server.internal.trainer.remote_downsamplers.irreducible_loss_producer import IrreducibleLossProducer
+
+
+class RemoteRHOLossDownsampling(AbstractRemoteDownsamplingStrategy):
+    """
+    Method adapted from Prioritized Training on Points that are Learnable, Worth Learning, and Not Yet Learnt
+    (Sören Mindermann+, 2022). https://arxiv.org/abs/2206.07137
+    """
+
+    def __init__(
+        self,
+        pipeline_id: int,
+        trigger_id: int,
+        batch_size: int,
+        params_from_selector: dict,
+        modyn_config: dict,
+        per_sample_loss: torch.nn.modules.loss,
+        device: str,
+    ) -> None:
+        super().__init__(pipeline_id, trigger_id, batch_size, params_from_selector, modyn_config, device)
+        self.per_sample_loss_fct = per_sample_loss
+        self.irreducible_loss_producer = IrreducibleLossProducer(
+            per_sample_loss,
+            modyn_config,
+            params_from_selector["rho_pipeline_id"],
+            params_from_selector["il_model_id"],
+            device,
+        )
+        self.rho_loss: torch.Tensor = torch.tensor([])
+        self.number_of_points_seen = 0
+
+    def init_downsampler(self) -> None:
+        self.index_sampleid_map: list[int] = []
+        self.rho_loss = torch.tensor([])
+        self.number_of_points_seen = 0
+
+    def inform_samples(
+        self,
+        sample_ids: list[int],
+        forward_input: Union[dict[str, torch.Tensor], torch.Tensor],
+        forward_output: torch.Tensor,
+        target: torch.Tensor,
+        embedding: Optional[torch.Tensor] = None,
+    ) -> None:
+        training_loss = self.per_sample_loss_fct(forward_output, target).detach()
+        self.index_sampleid_map += sample_ids
+        irreducible_loss = self.irreducible_loss_producer.get_irreducible_loss(sample_ids, forward_input, target)
+        self.rho_loss = torch.cat([self.rho_loss, training_loss - irreducible_loss]).to(training_loss.dtype)
+        self.number_of_points_seen += forward_output.shape[0]
+
+    def select_points(self) -> tuple[list[int], torch.Tensor]:
+        target_size = max(int(self.downsampling_ratio * self.number_of_points_seen / 100), 1)
+        # find the indices of maximal "target_size" elements in the list of rho_loss
+        selected_indices = torch.topk(self.rho_loss, target_size).indices
+        # use sorted() because we keep the relative order of the selected samples
+        selected_sample_ids = [self.index_sampleid_map[i] for i in sorted(selected_indices)]
+        # all selected samples have weight 1.0
+        selected_weights = torch.ones(target_size)
+        return selected_sample_ids, selected_weights
```

Comment:
Something I just realized is that the RHO-LOSS paper does Batch-then-sample, however it does not accumulate batches which seems odd to me. Let's stick with what we have for now, and then see how it performs.

Reply:
Reading this implementation, I wonder whether we can also allow sample then batch? I know the original paper does bts (without accumulation) but I don't see any reason why this would not work in a StB setting

Reply:
Implementation-wise it is not a problem at all to have sample-then-batch. It is just that we then select the entire epoch for use, and keep updating the model using this epoch. So the samples selected to use in the second half of this epoch might not be up-to-date.

We previously agreed on this in doc https://docs.google.com/document/d/19dAm9IfVz4tK8-y8tTuo4lIiy5P7jC-juJveyJYB1AI/edit#heading=h.nrwv8u52nq3j.

But sure we can also have it and do experiments, what do you think?

Reply:
Okay the conclusion after our discussion: we will keep it as-is for now. Later when we want to test sample-then-batch, there is no implementation block for us.